### PR TITLE
New classes for commands and instructions

### PR DESCRIPTION
**Pull request by benjvmin93**
Before submitting, please check the following:
- Make sure you have tests for the new code and that test passes (run `pytest`)
- If applicable, add a line to the [unreleased] part of CHANGELOG.md, following [keep-a-changelog](https://keepachangelog.com/en/1.0.0/).
- Format added code by `black` and `isort`
  - See `pyproject.toml` for configurations

Then, please fill in below:

**Context (if applicable):**

**Description of the change:**
Introduced `class Command` and `class Instruction` for readability, maintainability and data validation purposes.

These classes allow to get rid of the list representation of both circuit model instructions, and the pattern commands. They are implemented using pydantic, which allows to validate the datas passed to the classes during instanciation.

**Related issue:**

also see that checks (github actions) pass.
If lint check keeps failing, try installing black==22.8.0 as behavior seems to vary across versions.

### CODE REVIEW

File: CHANGELOG.md
Comment (thierry-martinez) [inline CHANGELOG.md:27]: You should rephrase the change in terms of API changes: the user cannot write `pattern.add(["M", 0, "XY", 0, [], []])` anymore but should write `pattern.add(M(node=0))` for instance.

File: benchmarks/statevec.py
Comment (thierry-martinez) [inline benchmarks/statevec.py:104]: This kind of changes are probably due to `black` not having been called with `-l 120`. They should not appear in the PR.

File: examples/noisy_mbqc.py
Comment (thierry-martinez) [inline examples/noisy_mbqc.py:21]: This line should not have been changed.

Comment (thierry-martinez) [reply examples/noisy_mbqc.py:21]: More generally, all the changes in this file seem to be only reformatting: maybe we should reformat this file, but it should be done in another PR.

File: examples/pattern_fragments.py
Comment (thierry-martinez) [inline examples/pattern_fragments.py:22]: This file should not have been changed.

File: examples/pattern_fragments.py
Comment (thierry-martinez) [inline examples/pattern_fragments.py:25]: We should keep `.pattern`.

File: graphix/pattern.py
Comment (thierry-martinez) [inline graphix/pattern.py:71]: If you add `from __future__ import annotations`, you may use `list[command.Command]`.

File: graphix/pattern.py
Comment (thierry-martinez) [inline graphix/pattern.py:197]: If you use `from __future__ import annotations`, you may get rid of quotation marks around `Pattern`.

File: graphix/pattern.py
Comment (thierry-martinez) [inline graphix/pattern.py:204]: It would be nicer to have `filter: list[command.CommandKind]`.

File: graphix/pattern.py
Comment (thierry-martinez) [inline graphix/pattern.py:459]: May we have `op: command.CommandKind`?

File: graphix/command.py
Comment (thierry-martinez) [inline graphix/command.py:8]: Can be written `Literal["XY", "YZ", "XZ"]`. But may we use `pauli.Plane` instead?

**Review by thierry-martinez — CHANGES_REQUESTED** [review]
Well done! The diff set is huge but I believe it is mainly the fault of `black` having been invoked without `-l 120`. I suggested some corrections to improve even more the use of declarative types.

File: examples/noisy_mbqc.py
Comment (benjvmin93) [reply examples/noisy_mbqc.py:21]: Is it really useful to make a whole PR for reformatting? I could maybe just revert the changes implied by the wrong use of black for the concerned files?

File: graphix/pattern.py
Comment (benjvmin93) [reply graphix/pattern.py:71]: I guess I will also have to echo this change to the other `list` type specifications, along with the tuple ones?

**Comment by shinich1** [comment]
@thierry-martinez could you resolve the above comments if think they're already taken care of (if not, please leave them open)?

File: graphix/gflow.py
Comment (shinich1) [inline graphix/gflow.py:1]: could you update `get_corrections_from_pattern`?

File: graphix/simulator.py
Comment (shinich1) [inline graphix/simulator.py:180]: let us wait until we deprecate (unsupport) <=py 3.9 to change to this.

File: graphix/transpiler.py
Comment (shinich1) [inline graphix/transpiler.py:978]: same as a above

File: graphix/visualization.py
Comment (shinich1) [inline graphix/visualization.py:309]: would `black` with `-l 120` option help to minimize diff?

File: graphix/command.py
Comment (shinich1) [inline graphix/command.py:1]: let us add to API reference: in `docs/source/generator.rst`, under the heading `Pattern Generation`, add like the following:

```
:mod:`graphix.command` module
++++++++++++++++++++++++++++++++

.. currentmodule:: graphix.command

.. autoclass:: N

[add others]
```

File: graphix/command.py
Comment (shinich1) [inline graphix/command.py:21]: make this abc

File: graphix/instruction.py
Comment (shinich1) [inline graphix/instruction.py:27]: make this abc

File: graphix/command.py
Comment (shinich1) [inline graphix/command.py:10]: (comment by @EarlMilktea)  - it does not make sense for this to inherit from `str` - e.g.  this should not inherit [string methods](https://docs.python.org/3/library/stdtypes.html#string-methods).

File: graphix/sim/statevec.py
Comment (shinich1) [inline graphix/sim/statevec.py:204]: please annotate with `npt.NDArray` instead of `np.ndarray`.

**Comment by shinich1** [comment]
great work @benjvmin93 and sorry for the delay reviewing this. I added various comments. 
Let us aim to merge this after #135 is merged (should be very soon).

File: graphix/clifford.py
Comment (EarlMilktea) [inline graphix/clifford.py:100]: I feel it's not readable.

How about removing trailing commas?

File: graphix/sim/statevec.py
Comment (EarlMilktea) [inline graphix/sim/statevec.py:2]: Please use `__futures__` instead.

File: graphix/sim/tensornet.py
Comment (EarlMilktea) [inline graphix/sim/tensornet.py:3]: Use `__future__`.

File: graphix/transpiler.py
Comment (EarlMilktea) [inline graphix/transpiler.py:11]: Use `__future__`.

File: graphix/transpiler.py
Comment (EarlMilktea) [inline graphix/transpiler.py:24]: Let's use either `pydantic.BaseModel` or `dataclass`, not both.

They have many in common.

File: graphix/transpiler.py
Comment (EarlMilktea) [inline graphix/transpiler.py:11]: Please import from `collections.abc` rather than `typing`.

File: tests/test_pattern.py
Comment (EarlMilktea) [inline tests/test_pattern.py:316]: Unnecessary `()`.

**Review by EarlMilktea — CHANGES_REQUESTED** [review]

File: graphix/pattern.py
Comment (shinich1) [inline graphix/pattern.py:382]: NOTE: the internal logic of `is_standard` has changed for performnce.

File: graphix/pattern.py
Comment (shinich1) [inline graphix/pattern.py:692]: NOTE: also logic changed for performance.

**Review by shinich1 — APPROVED** [review]
All comments are resolved, and I believe the changes look sound. 
A few changes (optimizations) of the internal logics were marked above for future reference.
I ask @EarlMilktea to take a quick final look before we merge this.
Thank you @benjvmin93 and @thierry-martinez for the great work!

**Comment by EarlMilktea** [comment]
@benjvmin93 @thierry-martinez 

Great job!

I will submit small PRs directly to your fork later.

**Comment by EarlMilktea** [comment]
MEMO: Usage of `abc.ABC` may need refactor later.

**Review by EarlMilktea — APPROVED** [review]

**Review by thierry-martinez — APPROVED** [review]

**Comment by thierry-martinez** [comment]
I think it is all green now: @shinich1, what do you think? Could @benjvmin93 merge the PR?

**Comment by shinich1** [comment]
@benjvmin93 @thierry-martinez great, please squash and merge.

**Pull request merged**
